thinc/optimizers.py

```diff
@@ -354,4 +375,201 @@ def _adam(self, ops, weights, gradient, lr_scale, key, nr_upd):
+class SWA(OptimizerABC):
+    """
+    After the optimization ran 'start_step' number of steps
+    it switches the learning-rate of the optimizer and starts
+    recording the moving averages of all weights.
+    """
+    optimizer: Optimizer
+    start_step: int
+    freq: int
+    learn_rate: FloatOrSeq
+    nr_update: Dict[Tuple[int, str], int]
+    run_swa: bool
+    averages: Dict[Tuple[int, str], FloatsXd]
+
+    def __init__(
+        self,
+        optimizer: Optimizer,
+        learn_rate: FloatOrSeqOrStr,
+        *,
+        freq: int = 1,
+        start_step: int = 0,
+        on_plateau: bool = False
+    ):
+        """
+        optimizer: The inner optimizer whose iterates to average.
+        learn_rate: Learning-rate (or schedule) to run with after
+            the averaging is turned on.
+        freq: Every 'freq' step the moving-averages will be updated.
+        start_step: Start the averaging after start_step number of updates.
+        """
+        self.optimizer = optimizer
+        self.start_step = start_step
+        self.freq = freq
+        self.learn_rate = learn_rate
+        self.nr_update = defaultdict(int)
+        self.run_swa = False
+        self.averages = {}
+        if isinstance(self.learn_rate, str):
+            if self.learn_rate != "constant":
+                raise ValueError(
+                    "The only string option that can be "
+                    "provided as learning-rate for SWA is"
+                    " 'constant' currently."
+                )
+
+    def _update_averages(
+        self,
+        key: Tuple[int, str],
+        weights: FloatsXd,
+    ):
+        """
+        Update moving average of weights.
+        """
+        if key not in self.averages:
+            ops = get_array_ops(weights)
+            self.averages[key] = ops.alloc(weights.shape, dtype="float32")
+            self.averages[key] += weights
+        else:
+            self.averages[key] *= self.nr_update[key]
+            self.averages[key] += weights
+            self.averages[key] /= self.nr_update[key] + 1
+
+    def _swap_lr(self):
+        """
+        Overwrite the optimizer's learning rate.
+        """
+        if self.learn_rate == "constant":
+            lr = self.optimizer.learn_rate
+            lr = constant(lr)
+        else:
+            lr = self.learn_rate
+        self.optimizer._set_attr_or_schedule("learn_rate", lr)
+
+    def __call__(
+        self,
+        key: Tuple[int, str],
+        weights: FloatsXd,
+        gradient: FloatsXd,
+        *,
+        lr_scale: float = 1.0,
+    ):
+        self.nr_update[key] += 1
+        nr_update = self.nr_update[key]
+        weights, gradient = self.optimizer(key, weights, gradient, lr_scale=lr_scale)
+        # Update running mean
+        if nr_update == self.start_step:
+            self._swap_lr()
+        if nr_update >= self.start_step and nr_update % self.freq == 0:
+            self._update_averages(key, weights)
```

Comment:
If I understand correctly, I think there is an ordering error here. `nr_update[key]` is incremented at the beginning of the function, but `_update_averages` assumes the old count?